### ISSUE TITLE
Support Object ACLs

### ISSUE THREAD

**Issue by rfornea** [minio/minio#8195]
I've read through a few github issues and the minio docs, and it seems you don't support setting object ACLs (nor intend to).  You claim to be an S3 compatible server, but if you aren't supporting this feature, that's not really true IMO.  Regardless of whether or not AWS recommends using ACLs, they nevertheless still support them and it is a good enough solution for plenty of people's usage.  Not supporting object ACLs prevents Minio from being a drop-in replacement for S3.  Someone who wanted to switch to Minio would have to start adding prefixes to everything which could break their existing naming scheme for objects.  This is a problem if they wanted to be able to support objects stored in S3 and in Minio, and if they already had pre-existing data in S3 which used Object ACLs.  This is the situation I personally am in.  

I really think you should reconsider this.

**Comment by harshavardhana** [comment]
ACLs are not implemented because they are incorrectly implemented on AWS S3, if you wish to spec compatible. Your application is better off not using it.

Most Enterprise applications that we came across such as Hadoop, Splunk, Apache NiFi, Kafka and a 100 others do not use ACLs. So there is no good reason for us to support ACLs.

**Comment by rfornea** [comment]
What is incorrect about them?

**Comment by harshavardhana** [comment]
ACLs are incorrectly implemented - let's take two scenarios here.

Scenario 1 - Bucket's having public-read and performance costs.

- enable BucketACL such as public-read on a bucket
- you still need to explicitly enable per object public-read by inheriting the bucketACL
  so this incurs a performance cost. Every Put now involves a secondary call.  Look at the alluxio
  implementation https://github.com/Alluxio/alluxio/blob/master/underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java#L545 

I added this code to turn this feature off in Alluxio because this secondary call has a performance cost and most of the time useless. It is simply implemented for historical reasons. This is a buggy implementation by AWS S3 to require applications to make a side-channel call to apply public-readiness of the object.  This side-channel is also not protected because anyone can change the bucketACL after the fact that you have read it. Since there is no exclusive lock involved between the call that you make to getBucketACL and actual creation of the object. 

MinIO can take thousands of calls per second, you can look at our performance numbers. Metadata heavy operations like ACLs can take a significant hit in these type of implementations and it is slow even with AWS S3 when the volume is significantly higher.

- This is a buggy implementation from AWS S3.

Scenario 2 - fully public read-write bucket where an anonymous user can upload an object
- Sets BucketACL as public-read-write
- Now upload an object to AWS S3 - using anonymous user. 
- Try accessing this object now as authenticated user - user sees Access Denied.
- Authenticated user already has access to the bucket, do not know why they would return Access Denied.
- This is a buggy implementation. 

That is why to better handle object-level authorization you are better off using bucket and IAM policies which is what AWS actively recommends now as ACLs are slowly being phased out.

**Comment by harshavardhana** [comment]
Our wish is to not bloat MinIO with two ways to do the same thing, ACLs were not implemented because as you can see they do not serve the required purpose and also require confusing API usage by applications. We would like to actively discourage such usage. Authorization is correctly and better handled by policies rather than ACLs and that is what we fully support.
 
I can go on listing the issues in the ACL implementation but I guess an explanation of the two scenarios is sufficient in this context.

**Comment by rfornea** [comment]
The performance issue seems like something that should be up to the discretion of the person who is running the server.  They might be fine with the performance hit.  

In your second scenario, being able to set the _object_ ACL should fix the issue.  

I don't really see it as a problem that objects don't automatically inherit the bucket ACL when the object can have its own ACL.  It's a cautious choice but it's not really "buggy."  Having new objects be private by default is erring on the side of safety.  

Object ACLs are nice because an object can be private in a public bucket or public in a private bucket, and if you needed to programmatically change the access, the URL to get to the object wouldn't change.

**Comment by harshavardhana** [comment]
I don't think you are able to see the problem in ACLs which is unfortunate. I rest my case.

**Comment by harshavardhana** [comment]
To close as a summary here - we will not implement ACLs @rfornea for any foreseeable future. The problems stated above are a major hindrance and unexpected design of ACLs cause complications when used in conjunction with Policies.  

Unless there is a serious application which cannot do without ACLs and there is just no way for it to migrate (I highly doubt that) we have no interest in implementing it. In most applications as explained and validated with MinIO ACLs are not used and newer applications which use AWS S3 are not using ACLs anymore due to "marked" as legacy API.

If you still want to use ACLs you should use alternative S3 compatible vendors which do implement ACLs at this point.

**Comment by tafel** [comment]
I was searching for a way to add ACLs to newly uploaded files and I found this page. 

I understand you don't want to add this feature, as it's not the best way to manage rights. But the storage I'm using, which is s3-compatible, still lacks support of bucket policies. So I really need canned ACLs.

For other people like me, I just want to share the solution I found for my use case.

It not documented anywhere, but it's just a simple line to add:

```js
const stream; // some stream from an upload
const size; // size of file
await minio.putObject('my-bucket', 'my/object.pdf', stream, size, {
    'x-amz-acl': 'public-read'
});
```
Last argument is metadata. And these metadatas are added as header to the request. And it's enough to add `x-amz-acl` header to let know the object storage you want an ACL here.

Anyway, thanks for you job at minio!
